crates/biome_js_analyze/src/lint/nursery/no_magic_numbers.rs

```diff
@@ -0,0 +1,433 @@
+/// Example: `const MyType: 100;` where `100` is the numeric literal type annotation.
+fn is_ts_numeric_literal_type_annotation(numeric_literal: &TsNumberLiteralType) -> bool {
+    get_sanitized_ts_parent_node(numeric_literal)
+        .is_some_and(|parent| TsTypeAnnotation::can_cast(parent.kind()))
+}
+
+/// Checks if the given `numeric_literal` is a numeric literal return type
+/// Example: `function f(x: string): 100 { return 100 }` where `100` is the numeric literal return type.
+fn is_ts_numeric_literal_return_type(numeric_literal: &TsNumberLiteralType) -> bool {
+    get_sanitized_ts_parent_node(numeric_literal)
+        .is_some_and(|parent| TsReturnTypeAnnotation::can_cast(parent.kind()))
+}
+
+/// Omits unary plus or minus expressions by returning the parent node if
+/// the current node is a unary expression with a plus or minus operator.
+/// Example: `-5` or `+3` will return the parent node, effectively skipping the unary operator parent node.
+fn omit_unary_plus_minus_parent(node: SyntaxNode<JsLanguage>) -> Option<SyntaxNode<JsLanguage>> {
+    let unary_expression_plus_or_minus =
+        JsUnaryExpression::cast(node.clone()).and_then(|unary_node| {
+            if unary_node.operator().is_ok_and(|operator| {
+                operator == JsUnaryOperator::Plus || operator == JsUnaryOperator::Minus
+            }) {
+                Some(unary_node)
+            } else {
+                None
+            }
+        });
+
+    if unary_expression_plus_or_minus.is_some() {
+        node.parent()
+    } else {
+        Some(node)
+    }
+}
+
+/// Omits the parenthesized expression if the node is wrapped in parentheses.
+/// Example: `(5)` will return the parent node, effectively skipping the parenthesized expression.
+fn omit_parenthesized_parent(node: SyntaxNode<JsLanguage>) -> Option<SyntaxNode<JsLanguage>> {
+    if JsParenthesizedExpression::can_cast(node.kind()) {
+        node.parent()
+    } else {
+        Some(node)
+    }
+}
+
+/// Returns the parent node of the given numeric literal, skipping over any unary plus/minus or parenthesized expression wrappers.
+/// It helps determine the true syntactic context of the literal by ignoring these common, but semantically insignificant, wrappers.
+fn get_sanitized_parent_node(
+    numeric_literal: &AnyJsNumericLiteral,
+) -> Option<SyntaxNode<JsLanguage>> {
+    numeric_literal
+        .syntax()
+        .parent()
+        .and_then(|parent| omit_unary_plus_minus_parent(parent).and_then(omit_parenthesized_parent))
+}
+
+/// Returns the parent node of the given numeric literal, skipping over any parenthesized expression for TsNumberLiteralType.
+fn get_sanitized_ts_parent_node(
+    numeric_literal: &TsNumberLiteralType,
+) -> Option<SyntaxNode<JsLanguage>> {
+    numeric_literal
+        .syntax()
+        .parent()
+        .and_then(|parent| omit_unary_plus_minus_parent(parent).and_then(omit_parenthesized_parent))
+}
```

Comment:
These methods are essentially the same, so you can unify them by taking the syntax node as argument instead.

```suggestion
/// Returns the parent node of the given node, skipping over any unary plus/minus or parenthesized expression wrappers.
/// It helps determine the true syntactic context of the node by ignoring these common, but semantically insignificant, wrappers.
fn get_sanitized_parent_node(node: JsSyntaxNode) -> Option<SyntaxNode<JsLanguage>> {
    node
        .parent()
        .and_then(|parent| omit_unary_plus_minus_parent(parent).and_then(omit_parenthesized_parent))
}
```

Reply:
been thinking about that, wanted some sort of union here, but becomes too verbose if I use union.

will use syntax node instead, thanks for the suggestion